src/clr-angular/forms/datepicker/date-input.spec.ts

```diff
@@ -0,0 +1,275 @@
+                    const testEl: HTMLInputElement = <HTMLInputElement>document.createElement("INPUT");
+                    testEl.value = "01/02/2015";
+
+                    expect(dateIOService.inputDate).toBeUndefined();
+
+                    context.clarityDirective.onValueChange(testEl);
+
+                    expect(dateIOService.inputDate).not.toBeUndefined();
+                    expect(dateIOService.inputDate).toBe("01/02/2015");
+                });
+
+                it("sets the inputDate to an empty string if the value of the input is invalid", () => {
+                    const testEl: HTMLInputElement = <HTMLInputElement>document.createElement("INPUT");
+                    testEl.value = "01/02/201";
+
+                    expect(dateIOService.inputDate).toBeUndefined();
+
+                    context.clarityDirective.onValueChange(testEl);
+
+                    expect(dateIOService.inputDate).not.toBeUndefined();
+                    expect(dateIOService.inputDate).toBe("");
+                });
+
+                it("sets the date on the IOService", () => {
+                    const date: Date = new Date();
+
+                    expect(dateIOService.date).toBeUndefined();
+
+                    context.clarityDirective.date = date;
+
+                    expect(dateIOService.date).not.toBeUndefined();
+                    expect(dateIOService.date.getDate()).toBe(date.getDate());
+                    expect(dateIOService.date.getMonth()).toBe(date.getMonth());
+                    expect(dateIOService.date.getFullYear()).toBe(date.getFullYear());
+
+                    const inValidDate: Date = new Date("Test");
+
+                    context.clarityDirective.date = inValidDate;
+                    expect(dateIOService.date).toBeNull();
+                });
+            });
+
+            describe("Host Bindings & Listeners", () => {
+                it("binds the placeholderText to the placeholder of the input", () => {
+                    expect(context.clarityElement.placeholder).toBe("MM/DD/YYYY");
+                });
+
+                it("binds the input type correctly", () => {
+                    expect(context.clarityElement.type).toBe("text");
+
+                    enabledService.fakeIsEnabled = false;
+                    context.detectChanges();
+
+                    expect(context.clarityElement.type).toBe("date");
+                });
+
+                it("listens to the input change events", () => {
+                    spyOn(context.clarityDirective, "onValueChange");
+
+                    const inputEl = context.fixture.debugElement.query(By.directive(ClrDateInput));
+                    inputEl.triggerEventHandler("change", inputEl);
+
+                    expect(context.clarityDirective.onValueChange).toHaveBeenCalled();
+                });
+            });
+        });
+
+        describe("Datepicker with ngModel", () => {
```

Comment:
As mention on the phone, please make sure you have the following unit tests:

- When the user types in the input element, both ngModel and the datepicker services end up with the correct value.
- When the app changes the ngModel value, both the input element and the datepicker services end up with the correct value.
- When the services update the selected date, both the input element and ngModel end up with the correct value.